clientv3/retry.go

```diff
@@ -54,14 +63,50 @@ func (c *Client) newRetryWrapper() retryRpcFunc {
 	}
 }
 
+func (c *Client) newRdonlyRetryWrapper() retryRpcFunc {
+	return func(rpcCtx context.Context, f rpcFunc) error {
+		for {
+			err := f(rpcCtx)
+			if err == nil {
+				return nil
+			}
+
+			// always stop retry on etcd errors other than invalid auth token
+			eErr := rpctypes.Error(err)
+			if etcdErr, ok := eErr.(rpctypes.EtcdError); ok {
+				if etcdErr.Code() == rpctypes.ErrInvalidAuthToken.(rpctypes.EtcdError).Code() {
+					gterr := c.getToken(rpcCtx)
+					if gterr != nil {
+						return err // return the original error for simplicity
+					}
+					continue
+
+				}
+				return err
+			}
+
+			return err
+		}
+	}
+}
+
 type retryKVClient struct {
 	pb.KVClient
-	retryf retryRpcFunc
+	retryf       retryRpcFunc
+	rdonlyRetryf retryRpcFunc
```

Comment:
Having a special function for reads seems forced and there's some code duplication-- `retryKVClient` was meant only for handling failfast/mutable operations. It's not meant to handle Range because the failfast=false policy works fine. I guess it should be renamed to something like `retryWriteKVClient`. Probably have instead separate wrapper types:

```go
type retryWriteKvClient struct { pb.KVClient; retryRpcFunc } // overrides put/txn/compact
// same as old retryKvClient

type retryKvClient struct { retryWriteKvClient } // overrides put/txn/compact/range; put/txn/compact inherited from retryWriteKvClient
func (kv *retryKvClient) Range(...) { ... }

func retryAuth(...) {
    for {
        err := f(rpcCtx)
        if rpctypes.Error(err) != ErrInvalidAuthToken { return err }
        if c.getToken() != nil { return err}
    }
}

retryWrites := &retryWriteKvClient{kv, retryWrite}
retryKv := &retryKvClient{retryWriteKvClient{retryWrites, retryAuth}}
```

That way the auth and failfast retry policies can be independent of each other.